core/arch/arm/kernel/ftrace.c

```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2019, Linaro Limited
+ */
+
+#include <kernel/ftrace.h>
+#include <kernel/thread.h>
+#include <mm/mobj.h>
+#include <optee_rpc_cmd.h>
+#include <printk.h>
+#include <string.h>
+#include <trace.h>
+
+#define MIN_FTRACE_BUF_SIZE	512
+#define MAX_HEADER_STRLEN	128
+
+void ta_fbuf_init(struct ta_head *ta_head, struct tee_ta_session *s)
+{
+	struct ftrace_buf *fbuf;
+	size_t fbuf_size = 0;
+	int count = 0;
+
+	if (ta_head->ftrace_buf_end > ta_head->ftrace_buf_start)
+		fbuf_size = ta_head->ftrace_buf_end - ta_head->ftrace_buf_start;
+
+	if (fbuf_size >= MIN_FTRACE_BUF_SIZE) {
+		fbuf = (struct ftrace_buf *)ta_head->ftrace_buf_start;
+		fbuf->head_off = sizeof(struct ftrace_buf);
+		count = snprintk((char *)fbuf + fbuf->head_off,
+				 MAX_HEADER_STRLEN,
+				 "Function graph for TA: %pUl @ %p\n",
+				 (void *)&s->ctx->uuid, (void *)ta_head);
+
+		fbuf->buf_off = fbuf->head_off + count;
+		fbuf->curr_size = 0;
+		fbuf->max_size = fbuf_size - sizeof(struct ftrace_buf) - count;
+
+		s->fbuf_paddr = virt_to_phys(fbuf);
+		s->fbuf_size = fbuf_size;
+	} else {
+		s->fbuf_paddr = 0;
+		s->fbuf_size = 0;
+	}
+}
+
+void ta_fbuf_dump(struct tee_ta_session *s)
+{
+	struct ftrace_buf *fbuf = NULL;
+	struct mobj *mobj;
+	char *va;
+	uint32_t ret = 0;
+	uint32_t dump_size = 0;
+
+	/* Return in case ftrace buffer is empty */
+	if (!s->fbuf_size)
+		return;
+
+	if (!core_mmu_add_mapping(MEM_AREA_TA_RAM, s->fbuf_paddr,
+				  s->fbuf_size)) {
```

Comment:
Maybe an issue when `CFG_WITH_PAGER=y`. maybe consider ftrace expect no pager, and postpone pager support to later (once defined where the ftrace buffer should be located: paged memory? less secure TA_RAM?).

Reply:
Agree, will add config check for ftrace expect no pager. As I am not sure how symbolize script will work with pager enabled.